libyul/optimiser/RedundantAssignEliminator.cpp

```diff
@@ -215,101 +81,79 @@ void RedundantAssignEliminator::operator()(Leave const&)
+			// Default-construct it in "Undecided" state if it does not yet exist.
+			m_stores[assignment->variableNames.front().name][&_statement];
 }
 
-void RedundantAssignEliminator::merge(TrackedAssignments& _target, TrackedAssignments&& _other)
+void RedundantAssignEliminator::shortcutNestedLoop(TrackedStores const& _zeroRuns)
 {
-	joinMap(_target, move(_other), [](
-		map<Assignment const*, State>& _assignmentHere,
-		map<Assignment const*, State>&& _assignmentThere
-	)
-	{
-		return joinMap(_assignmentHere, move(_assignmentThere), State::join);
-	});
+	// Shortcut to avoid horrible runtime:
+	// Change all assignments that were newly introduced in the for loop to "used".
+	// We do not have to do that with the "break" or "continue" paths, because
+	// they will be joined later anyway.
+	// TODO parallel traversal might be more efficient here.
+	for (auto& [variable, stores]: m_stores)
+		for (auto& assignment: stores)
+		{
+			auto zeroIt = _zeroRuns.find(variable);
+			if (zeroIt != _zeroRuns.end() && zeroIt->second.count(assignment.first))
+				continue;
+			assignment.second = State::Value::Used;
+		}
 }
 
-void RedundantAssignEliminator::merge(TrackedAssignments& _target, vector<TrackedAssignments>&& _source)
+void RedundantAssignEliminator::finalizeFunctionDefinition(FunctionDefinition const& _functionDefinition)
 {
-	for (TrackedAssignments& ts: _source)
-		merge(_target, move(ts));
-	_source.clear();
+	for (auto const& param: _functionDefinition.parameters)
+		finalize(param.name, State::Unused);
+	for (auto const& retParam: _functionDefinition.returnVariables)
+		finalize(retParam.name, State::Used);
 }
 
 void RedundantAssignEliminator::changeUndecidedTo(YulString _variable, RedundantAssignEliminator::State _newState)
 {
-	for (auto& assignment: m_assignments[_variable])
+	for (auto& assignment: m_stores[_variable])
 		if (assignment.second == State::Undecided)
 			assignment.second = _newState;
 }
 
 void RedundantAssignEliminator::finalize(YulString _variable, RedundantAssignEliminator::State _finalState)
 {
-	std::map<Assignment const*, State> assignments;
-	joinMap(assignments, std::move(m_assignments[_variable]), State::join);
-	m_assignments.erase(_variable);
+	std::map<Statement const*, State> stores;
+	util::joinMap(stores, std::move(m_stores[_variable]), State::join);
+	m_stores.erase(_variable);
 
 	for (auto& breakAssignments: m_forLoopInfo.pendingBreakStmts)
 	{
-		joinMap(assignments, std::move(breakAssignments[_variable]), State::join);
+		util::joinMap(stores, std::move(breakAssignments[_variable]), State::join);
 		breakAssignments.erase(_variable);
 	}
 	for (auto& continueAssignments: m_forLoopInfo.pendingContinueStmts)
 	{
-		joinMap(assignments, std::move(continueAssignments[_variable]), State::join);
+		util::joinMap(stores, std::move(continueAssignments[_variable]), State::join);
 		continueAssignments.erase(_variable);
 	}
 
-	for (auto const& assignment: assignments)
+	for (auto const& store: stores)
```

Comment:
```suggestion
	for (auto&& [statement, state]: stores)
```
?

Reply:
considered this, but there is another local variable called `state`...

Reply:
How can I enforce `statement` and `state` to be non-references? Is `auto [statement, state]` enough?

Reply:
The condition below can just be
```
if ((state == State::Unused || (state == State::Undecided && _finalState == State::Unused)) && ...
```
or the inner one could just be renamed... but also fine keeping it.

Forcing ``auto [statement, state]`` to be of value type is something I wouldn't try myself :-). (The problem there not so much being ``auto``, but the structured binding working via weird intermediate tuple types by spec ;-))